README.md
Ad_man
======

Ad_man is an advertising manager mountable engine.  It allows the user to upload and manage advertising banners, and provides an easy to use helper for inserting the banners on your website.  Also counts the number of times an ad has shown, and the number of times an advertisement has been clicked on.  The user must set a keyword for each ad.

Currently supports two image sizes for advertising: Leaderboard 728 x 90 and Banner 468 x 60. 
TODO: Will implement 9 different sizes as mentioned by google's adwords requirements.

Requires Paperclip 3.1, and requires ImageMagick be running on the server.

## Getting started

Ad_man works with Rails 3.2 onwards. You can add it to your Gemfile with:

```ruby
gem 'ad_man'
```

After you install Ad_man and add it to your Gemfile, you need to run the generator:

```console
rails generate ad_man:install
```

This will install the migrations and a few controller files, then just rake the database:
```console
rake db:migrate
```

See where Ad_man is located by running: 
```console
rake routes
```

In our case, we needed to move the line generated by the install in the routes file to our admin space.  Then add any before_filters to the controllers that were generated.

Make sure to change references to the application using the main_app helper:
```erb
<%= link_to "Home", main_app.root_path %>
<%= link_to "Advertising", main_app.ad_man_path %>
```

One final change may be necessary in order to preview an image before saving, and that is to add this line to the admin view:
```erb
<%= javascript_include_tag "ad_man/advertisements.js" %>
```

## Changing default values

If you need to change the default values set by ad_man, you will need to create an ad_man.rb initializer in your config/initializers folder.  

```ruby
# AdMan Default Values
AdMan.leaderboard_size = "728x90"
AdMan.banner_size = "468x60"
AdMan.max_image_size = 50 # 50Kb maximum image size

# These are configurable dimensions (currently leaderboard default)
AdMan.image_dimensions_width = 728
AdMan.image_dimensions_height = 90

# Configurable content type for advertising
AdMan.content_type = ["image/jpg", "image/bmp", "image/png", "image/gif", "image/jpeg"]

# Configurable max advertising for keyword
AdMan.max_count = 6
```

## Putting an advertisement in the view

To put your advertisement in the view just add:
```erb
<%= link_to_ad %>
```
Currently there are three parameters that can control which ads get shown on your page.  :keyword, :size, and :display_on_all_pages.  
```erb
<%= link_to_ad :keyword => "Your Keyword", :size => "leaderboard", :display_on_all_pages => false %>
```
## Authors
Written by [David Strand](http://www.github.com/wspyder) and [Tyler Hu](http://www.github.com/tylerhu)

## Copyright & Licensing

Copyright (c) 2012 [Kudelabs](http://kudelabs.com)

Released under MIT License. 


history.txt
    1  cd distros
    2  pwd
    3  hdiutil convert -format UDRW -o ~/Users/dstrand/Desktop/distros/Smeegol.i686-1.0.5-Build1.18.img /Users/dstrand/Desktop/distros/Smeegol.i686-1.0.5-Build1.18.iso 
    4  hdiutil
    5  hdiutil convert -format UDRW -o ~/Users/dstrand/Desktop/distros/Smeegol.i686-1.0.5-Build1.18.img /Users/dstrand/Desktop/distros/Smeegol.i686-1.0.5-Build1.18.iso 
    6  hdiutil --help
    7  hdiutil help
    8  hdiutil convert -format UDRW -o /Users/dstrand/Desktop/distros/Smeegol.i686-1.0.5-Build1.18.img /Users/dstrand/Desktop/distros/Smeegol.i686-1.0.5-Build1.18.iso 
    9  diskutil list
   10  diskutil unmountDisk /dev/disk1
   11  sudo dd if=/Users/dstrand/Desktop/distros/Smeegol.i686-1.0.5-Build1.18.img.dmg of=/dev/disk1 bs=1m
   12  diskutil eject /dev/disk1
   13  diskutil list
   14  ls
   15  hdiutil convert -format UDRW -o BT5R1-GNOME-32.img BT5R1-GNOME-32.iso 
   16  diskutil unmountDisk /dev/disk1
   17  dd help
   18  dd --help
   19  sudo dd --help
   20  sudo dd
   21  sudo dd if=/Users/dstrand/Desktop/distros/BT5R1-GNOME-32.img.dmg of=/dev/disk1 bs=2m
   22  sudo dd
   23  diskutil eject /dev/disk1
   24  ifconfig
   25  curl
   26  curl -L http://download.opensuse.org/repositories/Meego:/Netbook:/1.2/openSUSE_Factory
   27  diskutil list
   28  diskutil eject /dev/disk1
   29  sudo dd if=/Users/dstrand/Desktop/distros/meego-netbook-ia32-chrome-1.2.0.img of=/dev/disk1 bs=2m
   30  diskutil list
   31  sudo dd if=/Users/dstrand/Desktop/distros/meego-netbook-ia32-chrome-1.2.0.img of=/dev/disk1 bs=2m
   32  diskutil unmount /dev/disk1
   33  diskutil unmount /dev/disk1
   34  diskutil unmountDisk /dev/disk1
   35  sudo dd if=/Users/dstrand/Desktop/distros/meego-netbook-ia32-chrome-1.2.0.img of=/dev/disk1 bs=2m
   36  diskutil eject /dev/disk1
   37  cd /efi/refit
   38  ls
   39  ./enable-always.sh
   40  sudo vi /private/etc/hosts
   41  sudo nano /private/etc/hosts
   42  sudo nano /private/etc/hosts
   43  cd /etc
   44  ls
   45  vi hosts
   46  su
   47  su
   48  sudo
   49  sudo vi hosts
   50  sudo vi hosts
   51  ping google.com
   52  ifconfig
   53  if exit;  help; quit
   54  ifconfig
   55  ifconfig
   56  ifconfig
   57  ifconfig
   58  ifconfig
   59  ping baidu.com
   60  ping baidu.com
   61  defaults write com.apple.finder AppleShowAllFiles TRUE
   62  defaults write com.apple.finder AppleShowAllFiles FALSE
   63  killall Finder
   64  defaults write com.apple.finder AppleShowAllFiles FALSE
   65  killall Finder
   66  /Applications/Nikon\ Software/Camera\ Control\ Pro\ 2/Camera\ Control\ Pro.app/Contents/MacOS/Camera\ Control\ Pro ; exit;
   67  ifconfig
   68  top
   69  kill -SIGUSR1 28075
   70  sudo kill -SIGUSR128075
   71  sudo kill -SIGUSR1 28075
   72  while [ true ]; do killall -INFO dd; sleep 30; done
   73  sudo while [ true ]; do killall -INFO dd; sleep 30; done
   74  login root
   75  diskutil list
   76  diskutil unmountDisk /dev/disk1s1
   77  ls
   78  cd Downloads
   79  ls
   80  cd Desktop
   81  ls
   82  ls
   83  cd ..
   84  cd Desktop
   85  cd distros
   86  ls
   87  pwd
   88  sudo dd if=/Users/dstrand/Desktop/distros/Linpus-Lite1.7.1.0.02.iso of=/dev/disk1 bs=1m
   89  sudo dd if=/Users/dstrand/Desktop/distros/Linpus-Lite1.7.1.0.02.iso of=/dev/disk1 bs=1m
   90  diskutil list
   91  jigdo-lite
   92  jigdo-lite
   93  ls
   94  cd ..
   95  cd ..
   96  ls
   97  cd Down
   98  cd Downloads/
   99  ls
  100  ./debian-6
  101  jigdo 
  102  jigdo-lite
  103  jigdo-lite
  104  jigdo-lite
  105  jigdo-lite
  106  jigdo-lite
  107  ifconfig
  108  ping 169.254.241.1
  109  ping 169.254.1.1
  110  ping 169.254.0.1
  111  ping 169.254.241.1
  112  ftp strandbeer.com
  113* 
  114  irb
  115  clear
  116  ifconfig
  117  ping baidu.com
  118  ls
  119  mkdir Code
  120  cd Code/
  121  ls
  122  gem install rails
  123  ifconfig
  124  login
  125  login root
  126  ls -la /Volumes/UNTITLED/GAMES/
  127  cd ..
  128  ls */
  129  ls /*
  130  ls *
  131  pwd
  132  cd /Volumes/UNTITLED/
  133  ls */
  134  rm -r Recycled/
  135  rm -r System\ Volume\ Information/
  136  ls -l
  137  rm -r 95e2962e41a38db1552f34b89b62/
  138  rm msdownld.tmp/
  139  rm -r msdownld.tmp/
  140  ls
  141  ls -la /Volumes/UNTITLED/GAMES/
  142  ls -la /Volumes/UNTITLED/BACKUP/
  143  cd /Volumes/UNTITLED/BACKUP/Pictures/
  144  ls -la
  145  cd 2007
  146  ls -la
  147  cd ..
  148  cd 2008
  149  ls -la
  150  ls -a
  151  cd .config
  152  ls
  153  cd ..
  154  cd .DS_Store 
  155  cd /
  156  login root
  157  cat /etc/hosts
  158  sudo edit /etc/hosts
  159  su edit /etc/hosts
  160  su edit /etc/hosts
  161  login root
  162  ping 192.168.1.106
  163  ssh 192.168.1.106
  164  ssh 192.168.1.106 root
  165  ssh -u root 192.168.1.106
  166  su
  167  ifconfig
  168  ping 192.168.1.1
  169  tcpdump
  170  man tcpdump
  171  ifconfig
  172  tcpdump -i en1 >> tcp_test.txt
  173  su
  174  quit
  175  su 
  176  cd down
  177  cd Downloads/
  178  cd wacaw-04
  179  ls
  180  ./wacaw --help
  181  ./wacaw -L
  182  ./wacaw -n 10 --jpeg --VGA testimage
  183  nmap 
  184  telnet 192.168.1.1
  185  ssh 192.168.1.1
  186  ./wacaw --jpeg --VGA testimage
  187  cron
  188  cd ..
  189  wacaw
  190  cat /etc/crontab
  191  su -
  192  crontab -l
  193  ls *la
  194  ls -la
  195  pwd
  196  cd ..
  197  ls -la
  198  crontab -e
  199  crontab -l
  200  textedit crontab
  201  ls -la
  202  ls
  203  cat /etc/hosts
  204  ifconfig
  205  cat /etc/hosts
  206  ping strandbeer.com
  207  ping www.strandbeer.com
  208  ping baidu.com
  209  ping google.com
  210  ping godaddy.com
  211  ping strandbeer.com
  212  ifconfig 
  213  cd Library/
  214  cd Application\ Support/
  215  ls
  216  ls -la
  217  rm .settings
  218  ftp
  219  ftp goodnote.biz
  220  ftp goodnote.biz
  221  ftp goodnote.biz
  222  ping goodnote.biz
  223  ping 208.109.78.137
  224  ssh goodnote.biz
  225  ping goodnote.biz
  226  ssh goodnote.biz
  227  ping goodnote.biz

app/models/ad_man/advertisement.rb
module AdMan
  class Advertisement < ActiveRecord::Base
    attr_accessible :destination_url, :title, :ad_banner, :display_count,
      :priority, :start_date, :end_date, :click_count, :keyword_id
    validates_with AttachmentPresenceValidator, :attributes => :ad_banner
    belongs_to :keyword
    has_attached_file :ad_banner, :styles => {
      :leaderboard => (AdMan.leaderboard_size)?(AdMan.leaderboard_size):'728X90', 
      :banner => (AdMan.banner_size)?(AdMan.banner_size):'468X60'
    }
    validates_attachment :ad_banner, :presence => true, 
      :content_type => {
        :content_type => (AdMan.content_type)?(AdMan.content_type): ["image/jpg","image/bmp","image/png", "image/gif", "image/jpeg"]
      },
      :size => { :in => 0..((AdMan.max_image_size)?(AdMan.max_image_size):50).kilobytes }
    validates_presence_of :destination_url, :title, :keyword_id, :priority
    validates_uniqueness_of :title
    validate :image_dimensions, :on => :create
    after_initialize :init_priority

    def Advertisement.render_random_ad(keyword_id = nil)
      #		ads = Advertisement.find_all_by_keyword_id(keyword_id)
      if keyword_id.nil?
        ads = Advertisement.where("start_date <= ? AND end_date >= ?", Date.today, Date.today)
      else 
        ads = Advertisement.where("keyword_id = ? AND start_date <= ? AND end_date >= ? ", keyword_id, Date.today, Date.today)
      end
      if ads.present? 
        total_times = 1.0
        total_priority = 0.0
        ads.each { |advertisement| 
          total_times += advertisement.display_count
          total_priority += advertisement.priority
        }
        ad = ads[rand(ads.size)]
        while((ad.display_count / total_times) > (ad.priority / total_priority))
          ad = ads[rand(ads.size)]
        end
        ad
      end
    end

    private
    def image_dimensions
      temp_file = ad_banner.queued_for_write[:leaderboard] #get the file
      dimensions = Paperclip::Geometry.from_file(temp_file)        
      max_width = (AdMan.image_dimensions_width)?(AdMan.image_dimensions_width):728
      max_height = (AdMan.image_dimensions_height)?(AdMan.image_dimensions_height):90
      if (dimensions.width > max_width) || (dimensions.height > max_height)
        errors.add("banner_dimmensions",
                   "must be image size: #{max_width}X#{max_height}.")
      end
    end

    def init_priority
      self.priority ||= 1
    end
  end
end


